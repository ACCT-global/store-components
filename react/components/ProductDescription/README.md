# Product Description
Product Description is a canonical component that any VTEX app can import.

To import it into your code: 
```js
import ProductDescription from 'vtex.store-components/ProductDescription'
```

## Usage
You can use it in your code like a React component with the jsx tag: `<ProductDescription />`. 
```jsx
<ProductDescription specifications={product.specs} skuName={product.name}> 
   <span>{product.description}</span>
</ProductDescription>
```

### Passing classes to elements of the component
The `classes` and `rowClasses` props have the following structure:
```js
{
  classes: PropTypes.shape({
    root: PropTypes.string,
    description: PropTypes.shape({
      container: PropTypes.string,
      title: PropTypes.string,
      text: PropTypes.string
    }),
    table: PropTypes.shape({
      container: PropTypes.string,
      title: PropTypes.string,
      tableElement: PropTypes.string,
      tableBody: PropTypes.string
    }),
  }),
  rowClasses: PropTypes.shape({
    row: PropTypes.string,
    thName: PropTypes.string,
    tdValue: PropTypes.string
  })
}
```

The attributes of `classes` and `rowClasses` objects represent an element of the component. To understand better see the following example of how to pass classes to elements
```jsx
// ...
const classes = {
  root: 'ma2',
  description: {
    title: 'f4 b ttu mb3',
    text: 'measure-wide'
  },
  table: {
    container: 'mt6',
    title: 'f4 b ttu mb6-ns mb5-s',
    tableElement: 'w-100'
  }
}

const rowClasses = {
  thName: 'dtc-ns bn-s',
  tdValue: 'c-muted-2 db-s'
}

<ProductDescription
  skuName={skuName}
  classes={classes}
  rowClasses={rowClasses}
  description={description}
  specifications={specifications}
/>
```

| Prop name | Type | Description |
| --------- | ---- | ----------- |
| `children` | `Node!` | Children component which contains the product description |
| `specifications` | `Array` | Specifications that will be displayed on the table |
| `specifications[n].name` | `String` | Specification name |
| `specifications[n].value` | `String` | Specifications value |
| `classes` | `Object` | CSS classes to be applied in the elements of the component |
| `rowClasses` | `Object` | CSS classes to be applied in the elements of SpecificationRow component |
| `skuName` | `String` | Name of the SKU |

See an example at [Product Details](https://github.com/vtex-apps/product-details) app
