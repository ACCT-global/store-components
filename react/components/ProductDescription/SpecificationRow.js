import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

export default function SpecificationRow({ name, values, classes }) {

  return (
    <tr className={classNames('vtex-product-specifications__table-row', classes.row)}>
      <th className={classNames('vtex-product-specifications__specification-name', classes.thName)}>
        {name}
      </th>
      <td className={classNames('vtex-product-specifications__specification-values', classes.tdValue)}>
        {values}
      </td>
    </tr>
  )
}

SpecificationRow.defaultProps = {
  classes: {
    row: null,
    thName: null,
    tdValue: null
  }
}

SpecificationRow.propTypes = {
  classes: PropTypes.shape({
    row: PropTypes.string,
    thName: PropTypes.string,
    tdValue: PropTypes.string
  }),
  name: PropTypes.node,
  values: PropTypes.node,
}
