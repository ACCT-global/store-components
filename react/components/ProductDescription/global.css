.vtex-product-specifications__table {
  margin: 0 auto;
  border-collapse: collapse;
}

.vtex-product-specifications__table-row:nth-child(odd) {
  background-color: rgba(0, 0, 0, 0.05);
}
