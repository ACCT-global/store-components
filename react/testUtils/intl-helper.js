import React from 'react'
import * as reactTestingLibrary from 'react-testing-library'
import { IntlProvider } from 'react-intl'
import defaultStrings from '../../messages/en-US.json'
const Context = React.createContext({})

export const render = (node, options) => {
  const rendered = reactTestingLibrary.render(
    <IntlProvider messages={defaultStrings} locale="en-US">
      {node}
    </IntlProvider>,
    options
  )

  return {
    ...rendered,
    rerender: newUi =>
      render(newUi, {
        container: rendered.container,
        baseElement: rendered.baseElement,
      }),
  }
}

// re-export everything
export default reactTestingLibrary
