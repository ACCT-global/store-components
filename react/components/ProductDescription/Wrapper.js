import React, { useContext } from 'react'
import { ProductContext } from 'vtex.product-context'
import { path } from 'ramda'

import ProductDescription from './index'

const ProductDescriptionWrapper = props => {
  const valuesFromContext = useContext(ProductContext)
  const description =
    props.description != null
      ? props.description
      : path(['product', 'description'], valuesFromContext)

  return (
    <ProductDescription
      description={description}
      showShowMoreButton={props.showShowMoreButton}
    />
  )
}

ProductDescriptionWrapper.defaultProps = {
  showShowMoreButton: true,
}

export default ProductDescriptionWrapper
