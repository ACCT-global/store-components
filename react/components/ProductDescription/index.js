import React, { Component, Fragment } from 'react'
import PropTypes from 'prop-types'
import ContentLoader from 'react-content-loader'
import { FormattedMessage, injectIntl, intlShape } from 'react-intl'

import SpecificationRow from './SpecificationRow'

import './global.css'

/**
 * Product Description Component.
 * Render the description and technical specifications of a product
 */
class ProductDescription extends Component {
  render() {
    const { specifications, skuName, description } = this.props

    if (!description || !specifications) {
      return null
    }

    return (
      <div className="vtex-product-description ma2">
        <div className="f4 b ttu mb3">
          <FormattedMessage id="product-description.title" />
        </div>

        <span
          className="measure-wide"
          dangerouslySetInnerHTML={{ __html: description }}
        />

        {specifications.length > 0 && (
          <div className="vtex-product-specifications mt6">
            <div className="vtex-product-specifications__title f4 b ttu mb6-ns mb5-s">
              <FormattedMessage id="technicalspecifications.title" />
            </div>
            <table className="vtex-product-specifications__table w-100">
              <tbody>
                {skuName && (
                  <SpecificationRow name={`SKU ${skuName}`} />
                )}
                {specifications.map(specification => (
                  <SpecificationRow
                    key={specification.name}
                    name={specification.name}
                    values={specification.values.map((value, i) => (
                      <Fragment key={value}>
                        <span dangerouslySetInnerHTML={{ __html: value }} />{' '}
                        {i !== specification.values.length - 1 && <br />}
                      </Fragment>
                    ))}
                  />
                ))}
              </tbody>
            </table>
          </div>
        )}
      </div>
    )
  }
}

ProductDescription.defaultProps = {
  specifications: [],
}

ProductDescription.propTypes = {
  /** Product description string */
  description: PropTypes.string,
  /** Intl object to provides internationalization */
  intl: intlShape.isRequired,
  /** Specifications that will be displayed on the table */
  specifications: PropTypes.arrayOf(
    PropTypes.shape({
      /** Specification name */
      name: PropTypes.string.isRequired,
      /** Specifications value */
      values: PropTypes.arrayOf(PropTypes.string).isRequired,
    })
  ),
  /** Name of the current SKU */
  skuName: PropTypes.string,
}

export default injectIntl(ProductDescription)
