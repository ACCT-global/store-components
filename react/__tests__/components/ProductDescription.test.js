import React from 'react'
import { render } from 'intl-helper'
import ProductDescription from './../../ProductDescription'

describe('<ProductDescription />', () => {
  const renderComponent = customProps => {
    const props = {
      description: 'Test description',
      ...customProps,
    }
    return render(<ProductDescription {...props} />)
  }

  it('should be mounted', () => {
    const { asFragment } = renderComponent()
    expect(asFragment()).toBeDefined()
  })

  it('should match the snapshot with description', () => {
    const { asFragment } = renderComponent()
    expect(asFragment()).toMatchSnapshot()
  })
})
