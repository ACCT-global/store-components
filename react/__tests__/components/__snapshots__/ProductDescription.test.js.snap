// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductDescription /> should match the snapshot with description 1`] = `
<div
  class="productDescriptionContainer"
>
  <h2
    class="t-heading-5 mb5 mt0"
  >
    Product Description
  </h2>
  <div
    class="c-muted-1"
  >
    <div
      class="relative"
      style="transition:600ms ease-in-out;height:220px;overflow:hidden"
    >
      <div
        class="h-auto"
      >
        Test description
      </div>
      <div
        class="pointerEventsNone flex absolute bottom-0 w-100 h-100 flex-column justify-end"
      >
        <div
          class="fadeBottom w-100 h-50"
          style="transition:400ms ease-in-out"
        />
        <div
          class="pointerEventsAuto bg-base tc w-100"
        >
          <div
            class="c-action-primary t-action pointer ma5"
          >
            <span>
              Show more
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
